### PR TITLE
dont make alias for none value

### PR DESCRIPTION
**Pull request by eellison**
Don't make an alias value for a value that is known to be None. This was preventing constant propagation from running the `out is None` check in nn.functional.normalize, and thus preventing the if statement from being inlined.

### CODE REVIEW

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 1]
```diff
@@ -20,7 +20,7 @@ bool AliasDb::shouldAnnotate(const TypePtr& type) {
 // We only need to annotate values that either are mutable or could contain
 // mutable types.
 bool AliasDb::shouldAnnotate(const Value* v) {
-  return shouldAnnotate(v->type());
+  return !v->mustBeNone() && shouldAnnotate(v->type());
```
Comment (suo) [inline torch/csrc/jit/passes/alias_analysis.cpp]: Not sure this is entirely safe. `shouldAnnotate` is a type-level concept, while `mustBeNone()` is a graph-level concept. What's actually inhibiting the constant prop from running here?

Comment (eellison) [reply torch/csrc/jit/passes/alias_analysis.cpp]: how is it not safe ? mustBeNone guarantees that the value is non-aliasing. Sure, we could restrict ourselves to doing analysis at the type level, but then we are throwing away information. 

In the following (edited) graph
graph(%i0 : Double(*, *, *)):
  %1 : None = prim::Constant()
  %3 : Tensor? = prim::Constant()
  %7 : bool = aten::__is__(%3, %1)
  %ret : Tensor = prim::If(%7)
    block0():
      ....
      -> (%ret.1)
    block1():
      %out.2 : Tensor = prim::unchecked_unwrap_optional(%3)
      %ret.2 : Tensor = aten::div(%i0, %denom.2, %out.2)
      -> (%ret.2)
  return (%ret)

%3 aliases %out.2, which is written to in aten::div. Because %3 has writers aten::__is__ doesn't run

Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.cpp]: I mean that in other places we use `shouldAnnotate` to enforce invariants that may be violated if we add this condition. We should be careful whenever we change the meaning of a core concept.

I think it would be better to put this in `makePointerTo`—what we want is to use extra information about the graph to deduce the fact that two things can never alias if at least one of them is None. I think there's already a check there that does something similar

Comment (eellison) [reply torch/csrc/jit/passes/alias_analysis.cpp]: Now with the suggested change, the assertion `AT_ASSERT(elementMap_.count(v));` fails in the above graph with aten::div. I'm not sure there's an easy way of solving this.

Comment (eellison) [reply torch/csrc/jit/passes/alias_analysis.cpp]: The issue is if you have some graph like, 
```
if cond:
   a = x1
else:
   a = x2
write(a)
```
and x1 and x2 both end up being none, the type of a will still be ?Tensor, and when you write to it in register_write will expect an element to be made to it. So the option here is to relax the invariants in register_write, carry around extra state about which values must be None, or make the change at the api point (which is what i did). i think that's minimally invasive.

**Comment by eellison** [comment]
@suo can u look

**Review by suo — APPROVED** [review]
